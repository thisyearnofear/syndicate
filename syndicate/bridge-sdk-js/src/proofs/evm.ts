import { createMPT, createMerkleProof } from "@ethereumjs/mpt"
import { RLP } from "@ethereumjs/rlp"
import { MapDB, bigIntToHex } from "@ethereumjs/util"
import { ethers } from "ethers"
import type { BlockTag, Log, TransactionReceipt, TransactionReceiptParams } from "ethers"
import type { EVMChainKind } from "../clients"
import { ChainKind, type EvmProof } from "../types"

interface BlockHeader {
  parentHash: string
  sha3Uncles: string
  miner: string
  stateRoot: string
  transactionsRoot: string
  receiptsRoot: string
  logsBloom: string
  difficulty: string
  number: string
  gasLimit: string
  gasUsed: string
  timestamp: string
  extraData: string
  mixHash: string
  nonce: string
  baseFeePerGas: string
  withdrawalsRoot?: string
  blobGasUsed?: string
  excessBlobGas?: string
  parentBeaconBlockRoot?: string
  requestsHash?: string
}

const RPC_URLS: Record<EVMChainKind, string> = {
  [ChainKind.Eth]: "https://eth.llamarpc.com",
  [ChainKind.Base]: "https://mainnet.base.org",
  [ChainKind.Arb]: "https://arb1.arbitrum.io/rpc",
}

function getChainRpcUrl(chain: EVMChainKind): string {
  const url = RPC_URLS[chain]
  if (!url) {
    throw new Error(`No RPC URL configured for chain: ${chain}`)
  }
  return url
}

class ExtendedProvider extends ethers.JsonRpcProvider {
  async getBlockReceipts(blockTag: BlockTag) {
    const receipts = await this.send("eth_getBlockReceipts", [this._getBlockTag(blockTag)])
    const network = await this.getNetwork()
    return receipts.map((receipt: TransactionReceiptParams) =>
      this._wrapTransactionReceipt(receipt, network),
    )
  }

  async getBlockHeader(blockTag: BlockTag): Promise<BlockHeader> {
    return this.send("eth_getBlockByNumber", [this._getBlockTag(blockTag), false])
  }
}

export async function getEvmProof(
  txHash: string,
  topic: string,
  chain: EVMChainKind = ChainKind.Eth,
): Promise<EvmProof> {
  const rpcUrl = getChainRpcUrl(chain)

  const provider = new ExtendedProvider(rpcUrl)
  const receipt = await provider.getTransactionReceipt(txHash)
  if (!receipt) {
    throw new Error(`Transaction receipt not found on ${ChainKind[chain]}`)
  }

  const blockNumber = bigIntToHex(BigInt(receipt.blockNumber))
  const [blockHeader, blockReceipts] = await Promise.all([
    provider.getBlockHeader(blockNumber),
    provider.getBlockReceipts(blockNumber),
  ])

  const { merkleProof, receiptData } = await buildReceiptProof(receipt, blockReceipts)
  const logData = findAndEncodeLog(receipt, topic)

  return {
    log_index: BigInt(logData.index),
    log_entry_data: logData.encoded,
    receipt_index: BigInt(receipt.index),
    receipt_data: receiptData,
    header_data: encodeBlockHeader(blockHeader),
    proof: merkleProof,
  }
}

async function buildReceiptProof(receipt: TransactionReceipt, blockReceipts: TransactionReceipt[]) {
  const trie = await createMPT({ db: new MapDB() })

  await Promise.all(
    blockReceipts.map(async (r) => {
      if (!r) throw new Error("Invalid receipt in block")
      const receiptRlp = encodeReceipt(r)
      const key = RLP.encode(r.index)
      await trie.put(key, receiptRlp)
    }),
  )

  const receiptKey = RLP.encode(receipt.index)
  const merkleProof = await createMerkleProof(trie, receiptKey)
  const receiptData = encodeReceipt(receipt)

  return { merkleProof, receiptData }
}

function findAndEncodeLog(receipt: TransactionReceipt, topic: string) {
  const logEntry = receipt.logs.find((log) => log.topics[0] === topic)
  if (!logEntry) {
    throw new Error("Log entry not found for the given topic")
  }

  return {
    index: receipt.logs.indexOf(logEntry),
    encoded: encodeLog(logEntry),
  }
}

function encodeReceipt(receipt: TransactionReceipt): Uint8Array {
  const items = [
    receipt.status ? "0x1" : "0x",
    receipt.cumulativeGasUsed,
    receipt.logsBloom,
    receipt.logs.map((log) => [log.address, Array.from(log.topics), log.data]),
  ]

  if (receipt.type === 0) {
    return RLP.encode(items)
  }

  return new Uint8Array([receipt.type, ...RLP.encode(items)])
}

function encodeLog(log: Log): Uint8Array {
  return RLP.encode([log.address, Array.from(log.topics), log.data])
}

function encodeBlockHeader(header: BlockHeader): Uint8Array {
  const items = [
    header.parentHash,
    header.sha3Uncles,
    header.miner,
    header.stateRoot,
    header.transactionsRoot,
    header.receiptsRoot,
    header.logsBloom,
    header.difficulty,
    header.number,
    header.gasLimit,
    header.gasUsed,
    header.timestamp,
    header.extraData,
    header.mixHash,
    header.nonce,
    header.baseFeePerGas,
    header.withdrawalsRoot,
    header.blobGasUsed,
    header.excessBlobGas,
    header.parentBeaconBlockRoot,
    header.requestsHash,
  ]
    .filter((item) => item !== undefined)
    .map((item) => (item === "0x0" ? "0x" : item))

  return RLP.encode(items)
}
