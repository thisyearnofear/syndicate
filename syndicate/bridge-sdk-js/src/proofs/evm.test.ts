import type { WithImplicitCoercion } from "node:buffer"
import { describe, it } from "vitest"
import { ChainKind, type EvmProof } from "../types"
import { getEvmProof } from "./evm"

describe.concurrent("ProofGenerator", () => {
  const ERC20_TRANSFER_TOPIC = "0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef"

  const formatProof = (proof: EvmProof) => ({
    log_index: proof.log_index.toString(),
    receipt_index: proof.receipt_index.toString(),
    receipt_data: Buffer.from(proof.receipt_data).toString("hex"),
    log_entry_data: Buffer.from(proof.log_entry_data).toString("hex"),
    header_data: Buffer.from(proof.header_data).toString("hex"),
    proof: proof.proof.map((p: WithImplicitCoercion<ArrayBuffer | SharedArrayBuffer>) =>
      Buffer.from(p).toString("hex"),
    ),
  })

  it("should generate proof for pre-Shapella transaction", async ({ expect }) => {
    const txHash = "0xc4a6c5cde1d243b26b013f805f71f6de91536f66c993abfee746f373203b68cc"

    const proof = await getEvmProof(txHash, ERC20_TRANSFER_TOPIC, ChainKind.Eth)
    expect(formatProof(proof)).toMatchSnapshot()
  })

  it("should generate proof for post-Shapella transaction", async ({ expect }) => {
    const txHash = "0xd6ae351d6946f98c4b63589e2154db668e703e8c09fbd4e5c6807b5d356453c3"

    const proof = await getEvmProof(txHash, ERC20_TRANSFER_TOPIC, ChainKind.Eth)
    expect(formatProof(proof)).toMatchSnapshot()
  })

  it("should generate proof for post-Dencun transaction", async ({ expect }) => {
    const txHash = "0x42639810a1238a76ca947b848f5b88a854ac36471d1c4f6a15631393790f89af"

    const proof = await getEvmProof(txHash, ERC20_TRANSFER_TOPIC, ChainKind.Eth)
    expect(formatProof(proof)).toMatchSnapshot()
  })

  it("should throw error for non-existent transaction", async ({ expect }) => {
    const txHash = "0x0000000000000000000000000000000000000000000000000000000000000000"

    await expect(getEvmProof(txHash, ERC20_TRANSFER_TOPIC, ChainKind.Eth)).rejects.toThrow(
      "Transaction receipt not found on Eth",
    )
  })

  it("should throw error for invalid topic", async ({ expect }) => {
    const txHash = "0xc4a6c5cde1d243b26b013f805f71f6de91536f66c993abfee746f373203b68cc"
    const invalidTopic = "0x0000000000000000000000000000000000000000000000000000000000000000"

    await expect(getEvmProof(txHash, invalidTopic, ChainKind.Eth)).rejects.toThrow(
      "Log entry not found for the given topic",
    )
  })

  it("should throw error for unsupported network", async ({ expect }) => {
    const txHash = "0xc4a6c5cde1d243b26b013f805f71f6de91536f66c993abfee746f373203b68cc"
    const topic = ERC20_TRANSFER_TOPIC
    // biome-ignore lint/suspicious/noExplicitAny: Testing purposes
    const chain = "invalid" as any

    await expect(getEvmProof(txHash, topic, chain)).rejects.toThrow(
      `No RPC URL configured for chain: ${chain}`,
    )
  })
})
