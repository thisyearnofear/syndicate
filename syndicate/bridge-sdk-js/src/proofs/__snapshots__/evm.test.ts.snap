// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProofGenerator > should generate proof for post-Dencun transaction 1`] = `
{
  "header_data": "f90258a03d83db510dbb1b68aab334d5effa70937bf609774f5ac2f50aa509eaee359f53a01dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d493479495222290dd7278aa3ddd389cc1e1d165cc4bafe5a08582aae9f1bfa5f49e356409b47aa002efc13d3876aaaeb15b75cc9e67634f7ca08b9f74f9c1ee8d442963a9067c3012e125aa9c8ede94803b049923dc14413e7ba00210aaa84a6a5f3cb493d89a3ed7e7bada07ba6475a5128a633d499ade487057b901001ca3141879f7f0191f8d0de3c8073a91148ea99119e988c099239b1dbe356b2488b1d923119100c56711db0b4f1231557a37e6689fb2bfcf476ffcdffcfeb2055c428c29b1b55f9d2d5fe1b9ab73793faa82107fde4c6f5906afc6f1e96fd0cd5b4b1e0de278c1aa99dcf8c8464b6addc396cc349d35cdb4c9db941f9a9ba6786b9593566201951f2d6ef57109025326f46149559b7277e9d4727cca4fd884f702b49130b726ea6287af63d828450ef3550f6758a105cb0c7c6f99c2cc0890f12d139ac74a9dec5a6b9a2e115b5750aa615c0f1bb502c895a7e751e3b8d266614bf1b1ca891780f1f01fdd8d72370d571cef98537dfaa5e14523c22521cb57bf8084012914cf8401c9c38083e2ef828465f9a74f8f6265617665726275696c642e6f7267a024743f84dbc9bf09324738eb9f72602197340e024d591f5bc532e017e14a5b8e880000000000000000850b0d2bfc6ca08e6ba4593486cd93f2099956109e915caa4dd15e32fe795530bd9c41840a7e79830c000080a05484c4252a6916e15e0c26f374b447ada8081da44f65dfa1768872a87f09798a",
  "log_entry_data": "f89b94438e48ed4ce6beecf503d43b9dbd3c30d516e7fdf863a0ddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3efa00000000000000000000000006889de313803839ce3e15996a86167f639b4a634a000000000000000000000000023ddd3e3692d1861ed57ede224608875809e127fa000000000000000000000000000000000000000000000001a2738b6f81a2e0130",
  "log_index": "0",
  "proof": [
    "f90131a0ea31efc223a1f91c0211cc8216c96d841a4c6aa9e706b1c80f4556473766c835a0a0d053a80857fe23752abc931238dfb519c600ec2c89740a8c0027f2452691dba07211c9aec8f2476236ff4ae404ea0a4d86a3b69aef1d7d68b4525b695d6937c3a066ff6c2bdfbec1a598afe01a4c8cb5342f09dc23e126af6161fc8b66a29258b4a02b1c9c12e45087ab52a964dac390a2b297b279839a14d73e6039ff4136efe176a0c5d3039faed9aefe6b036979ee41e741a6d1a1ffb280d245eefac5646ff0e5b4a0de9817fcc57347a50f1c3bf6bc0af45514a1e17b349d6b3a893acdfc45ad0a70a0bd3ac3599d9a58447c23c89c2ba279fa00a11c13c5e181012d50e012ef845f77a026ce9a58a9d343be345947cd441816397a516ad88e7211170a856d8ce164a14b8080808080808080",
    "f851a0776031d2861aa23cb9bd323e49ac2b831658c311c35413899369ea8d558af735a08b9654ecaae433ee01704bb66b4877076a1f8e40b92ba31cadfb9dc06f858c15808080808080808080808080808080",
    "f8f18080808080808080a07a49b3a8c05d073eb6d6d11c5c4fdc4505acb5a36c9cbc590929aad029ffc719a099264b31f39abaef84c246bbac4a72dc8bcf5d935c775a8d6c356234d0226061a042f9385fe9d42d8764f8e69b30cae57bcff98c77dc5431d7cc21d10905abcd1ca08a3023498e1067ee67f3caa5a290502efcb05549dd9b0814023e6e1448288467a06e40e4f27b3a0ed5de67c684369aa12610652e6399d002edb5443394f64db861a069631286f3eddfa8ef89eec490348240bc8690b874db62dde1690206dff758b2a0cebfcf4bb26fdcd28a9a42ec0da1e287a0cb41d9bb40ca4879d1cccaf23309998080",
    "f90211a0c36f6abaee0178f644673e7d81132cdb139a9bd0bacf0d9719803242b54ef4d3a001a13437394f2ec24ec6a62baf5ef685dc617948314cf6cb7cb237dd759ba7afa0723df3efb2a08ca7156749e43c789cc92001953180a0e51935b3db6eb9c189b2a0d9929bc3195b7cb531249cfcccb0d7d56243a3af67be40d8a89a1ee287a2deeaa0641a7b318fd965ea4d1eb0686909dcd7898b491bb632eeb1ebad2bcc828f4017a0908899d902b0d7d459bba5e19767fc3373764fc6290de12d6588b4e90694aadfa0c6bdef3e0259812cbb979859c0e6a5e2bf980be306d5ed709ae373f0953449cca00993685867da25fb8ac53d636b9630a20297cdabf6e45bc003770dc335742d99a0c8b548c83f65e60afe207d499697f7b377be59c73183f0507862bc739be2348ba0cb4bc9bafef0aec55cd9ff83bc459d162e9f3701b784e2af2a768bc8b3dceeb9a066f522fef288d829e8e49738af851f7aef120dda1f6a1b934c24f96f52684ff2a0117847f1836bf5f354706fce391a183f4b69a55c6e64e820d69c4f5ee35e703da0548eb3f38b406cd35381c28b8a78028d48fe63db72f4a595893adc579b566295a0228c5276786aad75391a7b42d96df14b4494723fc1b096e07bc73e6bfaebc00ea01ae3c79fc12a94557c83e2668c89ec273b0fec08ad6b66e758cde410f121d340a06009085a2fc2a80ee20e8846c9320e28fd7ecf2c519fb8b5559fa30c750ebe8b80",
    "f902ae20b902aa02f902a60183cee136b9010000000000000020000000000000000000000000000000000000000000000000000800010000000000000008000100000040000000000000000000000000000000000000000000000000000008004000000000000000000000000000000000000000000000000000000000000000000200000000000000010000000010000000000000010000000000000000000000000000000000000004000000000000000000000000000000000080000000200000000000000000000000000000004000000000000002000000000000000000000000000000000000080000210000000000000000000000000000000000000200000000000000000000000000000000000000f9019bf89b94438e48ed4ce6beecf503d43b9dbd3c30d516e7fdf863a0ddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3efa00000000000000000000000006889de313803839ce3e15996a86167f639b4a634a000000000000000000000000023ddd3e3692d1861ed57ede224608875809e127fa000000000000000000000000000000000000000000000001a2738b6f81a2e0130f8fc9423ddd3e3692d1861ed57ede224608875809e127ff863a0dd85dc56b5b4da387bf69c28ec19b1d66e793e0d51b567882fa31dc50bbd32c5a0000000000000000000000000438e48ed4ce6beecf503d43b9dbd3c30d516e7fda00000000000000000000000006889de313803839ce3e15996a86167f639b4a634b88000000000000000000000000000000000000000000000001a2738b6f81a2e01300000000000000000000000000000000000000000000000000000000000000040000000000000000000000000000000000000000000000000000000000000000b74696d626c652e6e656172000000000000000000000000000000000000000000",
  ],
  "receipt_data": "02f902a60183cee136b9010000000000000020000000000000000000000000000000000000000000000000000800010000000000000008000100000040000000000000000000000000000000000000000000000000000008004000000000000000000000000000000000000000000000000000000000000000000200000000000000010000000010000000000000010000000000000000000000000000000000000004000000000000000000000000000000000080000000200000000000000000000000000000004000000000000002000000000000000000000000000000000000080000210000000000000000000000000000000000000200000000000000000000000000000000000000f9019bf89b94438e48ed4ce6beecf503d43b9dbd3c30d516e7fdf863a0ddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3efa00000000000000000000000006889de313803839ce3e15996a86167f639b4a634a000000000000000000000000023ddd3e3692d1861ed57ede224608875809e127fa000000000000000000000000000000000000000000000001a2738b6f81a2e0130f8fc9423ddd3e3692d1861ed57ede224608875809e127ff863a0dd85dc56b5b4da387bf69c28ec19b1d66e793e0d51b567882fa31dc50bbd32c5a0000000000000000000000000438e48ed4ce6beecf503d43b9dbd3c30d516e7fda00000000000000000000000006889de313803839ce3e15996a86167f639b4a634b88000000000000000000000000000000000000000000000001a2738b6f81a2e01300000000000000000000000000000000000000000000000000000000000000040000000000000000000000000000000000000000000000000000000000000000b74696d626c652e6e656172000000000000000000000000000000000000000000",
  "receipt_index": "212",
}
`;

exports[`ProofGenerator > should generate proof for post-Shapella transaction 1`] = `
{
  "header_data": "f90242a038bfb1f6aebbce44036875c8fdb502be76544ea6ce013f3246cd0e2de23e12d2a01dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d4934794dafea492d9c6733ae3d56b7ed1adb60692c98bc5a0c437bf5d5172d185d95e2efdf08a025e1c823ca74f12807810e834b19d3dca53a061f71d063dbc7c40a1b6b58bb0cdd17b4393c18b8f0a73a3a38aa2b2553f97bda09d930fa6d75345f69209347167bf090322fb953121884604e57d520737e3ebf1b901003fe505c9682b67463f465224f764c7d919593c043e882d4d18fb124ea96ef5289e429fce830280aefa1b3f05e71953833b0394ffad9169619eb685ab1ffe59eba04ad458cd288a19bc4a476ed15d19a680817db883f48e4455ee3e18a2f14b14dee4e85ae63b06da3c3b6c8cb63878b106200f715849e72ade3592d8341b5c70d4c7967aba424ab9b27b93691ee98f36a9535f099b6034c9e4e42357ddd60e10132a85802992ff0a84b42aec3896df121e93dfcdf2aa385229491e2a47aed338c9923cca95c7484311d8dfc4c27f2811143e233bc7e7ed1d042cb947b4e4f4529375fd5a99c88922824f23a50a854c1aac35301ae7f84260c77d4a76e9e81fce8084012393cc8401c9c38083cf72e28465b6fa9f9f496c6c756d696e61746520446d6f63726174697a6520447374726962757465a0e55536c717d3026592d3e2f41b4f47abbcaf120817d9e5c175520508a6ce916888000000000000000085021447317fa0b1599e25a6d29e39783ce2856aa95451bc3ce66f924d5396e1e4b4ab8205f9ad",
  "log_entry_data": "f89b94a0b86991c6218b36c1d19d4a2e9eb0ce3606eb48f863a0ddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3efa000000000000000000000000090b0a46b1c02bd137ca1410772eff1c231f07fe2a000000000000000000000000023ddd3e3692d1861ed57ede224608875809e127fa00000000000000000000000000000000000000000000000000000000014dc9380",
  "log_index": "0",
  "proof": [
    "f90111a09e0259b98fa182a95c48a8dacc2692badd7c29f71bc3e0d9d73dbf9d07e9611ba0ba6c9f95a62c8f2b828d4160e01162934297442c7cefea304908609de9fbd075a0993a2a24f0772d90700a7ad8f133ab1b7963defa359ad9826eb963f6168cfb60a0c31053f554e06e64bdddc105051dd2210d396514cd6edcc9aa55767e7652825ba05e2a24f235c03c6b6320a1d62e0e6d646349da863767ecf10126e6b4aacf274fa0e0c4a45d2ecd16a37af707062a5bec9f6657fcfe14ddb692215bdd6a26600617a0e53fe123d45561809c69c532b39d90e7eb0b6d7a74d09a259cbb41a29b12081480a044cf8e4f60cf97fb129824b1b37c04e82251da90b5e8fb9b3998cc3519e159798080808080808080",
    "f90211a0d4de7e86f5e6eba4a08c89925937986e201e6f286b7123fb5700354df9167b7ba07b3389484104ebb32e6959db5ed3cddcdb2d1ce0c2b41ce5dc4ce2553e808f84a0e7344755384ea053c57c5546c1516a356c8cab5fe926a5a3eb3fe6721671001da0b9312cb00a3723a9a80598ab13d730430c686e9f2b14e669f41d87f36ede87b2a06069a92ac34df54f389475c9b92fea2b92bc73e92bffa14cb5c085469e884398a03f6262719ad775eef95b991f24aa63370892d4d0ef8dbaeba19f681e2cba2b5aa0071ab0cc93209db0be2343111f91ba35a180e0a453f23660aed462fce5daf551a0bfd7126fa52bb94255b9b7613a5a0b8bdb1f01443de5da83ce1839a0fe9d6e3ba02384631620ba5abce1a52e7347bff8c3727a56390bd7bc57c7d1386508cb1672a06ed35b57e37f3a08788866524b8a2d0089ec3fb308f34dea583324afd57e476aa05af302b741e36f6c630f423ff8d40be2fdeb4fb517eb62782ae6955c986ce21fa0b16f820757bb0bd971bdba9371540c09528d2818a6ab196ca6cbb3919d5097bda027b6b7c488fd95227a11f4c62d1d138341fd242f558de9bba659f99f543edd18a06f83a49cb804e10ee72d2cbd844b0fb6145eb5ca2f3bcd0c4d37fd357f2404d8a0b9f13ed67f0b534b781693633d36bde8f6127aa4cae31659475910a08a2956f4a0dcc805b1d26db82e140bf5de19cb39b95e7590a52e13cd83a65471cb95f456fd80",
    "f902cf20b902cb02f902c701837036edb9010000000000000000000000000000000000000000000000000000200000000000000800010000000000000002000100000000000000000000000000000000000000000000000000000008000008000000000000000000000000000000000000000000000000000000000000000000000000000000000000010000000010000000000000000000000000000000000000000200000000010000000000000000000000000000000000200080000000200000000000000000000000000000004000000000000002000000000000000000000000000000000000080000000000000000000000000000000000000000000200000004000000000002000000000000000000f901bcf89b94a0b86991c6218b36c1d19d4a2e9eb0ce3606eb48f863a0ddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3efa000000000000000000000000090b0a46b1c02bd137ca1410772eff1c231f07fe2a000000000000000000000000023ddd3e3692d1861ed57ede224608875809e127fa00000000000000000000000000000000000000000000000000000000014dc9380f9011c9423ddd3e3692d1861ed57ede224608875809e127ff863a0dd85dc56b5b4da387bf69c28ec19b1d66e793e0d51b567882fa31dc50bbd32c5a0000000000000000000000000a0b86991c6218b36c1d19d4a2e9eb0ce3606eb48a000000000000000000000000090b0a46b1c02bd137ca1410772eff1c231f07fe2b8a00000000000000000000000000000000000000000000000000000000014dc93800000000000000000000000000000000000000000000000000000000000000040000000000000000000000000000000000000000000000000000000000000002f6175726f72613a316339396663333736353537366635613641353544353341653431423936434239373141393432350000000000000000000000000000000000",
  ],
  "receipt_data": "02f902c701837036edb9010000000000000000000000000000000000000000000000000000200000000000000800010000000000000002000100000000000000000000000000000000000000000000000000000008000008000000000000000000000000000000000000000000000000000000000000000000000000000000000000010000000010000000000000000000000000000000000000000200000000010000000000000000000000000000000000200080000000200000000000000000000000000000004000000000000002000000000000000000000000000000000000080000000000000000000000000000000000000000000200000004000000000002000000000000000000f901bcf89b94a0b86991c6218b36c1d19d4a2e9eb0ce3606eb48f863a0ddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3efa000000000000000000000000090b0a46b1c02bd137ca1410772eff1c231f07fe2a000000000000000000000000023ddd3e3692d1861ed57ede224608875809e127fa00000000000000000000000000000000000000000000000000000000014dc9380f9011c9423ddd3e3692d1861ed57ede224608875809e127ff863a0dd85dc56b5b4da387bf69c28ec19b1d66e793e0d51b567882fa31dc50bbd32c5a0000000000000000000000000a0b86991c6218b36c1d19d4a2e9eb0ce3606eb48a000000000000000000000000090b0a46b1c02bd137ca1410772eff1c231f07fe2b8a00000000000000000000000000000000000000000000000000000000014dc93800000000000000000000000000000000000000000000000000000000000000040000000000000000000000000000000000000000000000000000000000000002f6175726f72613a316339396663333736353537366635613641353544353341653431423936434239373141393432350000000000000000000000000000000000",
  "receipt_index": "54",
}
`;

exports[`ProofGenerator > should generate proof for pre-Shapella transaction 1`] = `
{
  "header_data": "f90221a0b0ef4977fdd02f1c20199405ce33ab46a7e0e54af8fb91d057b15ae6ce2a2fe4a01dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d4934794388c818ca8b9251b393131c08a736a67ccb19297a06566509d5c745b77971539d99f6922b8590f77452abd2eed58e55a65e9c1d623a0246428114de35f083dab2c452c0a8c14cba39f5acca3dca689a1b81a32f4f400a063816472b1ef14461da67b2b8be8b3d9c2957c376e4114fa4aabeb2bbde29a69b90100bbf04f294913112e54b4021c81091e2911d14e8b35183522c2490c986c0611d03c448b640648840040a6d6d859c71b098e2721ec5ac4adad0a270c30d532b416c83f59b24902070d6de63c8955d100669004e9f06acc820002e04c00c9c1572a1b5084584e2a540503a2d4c1c8120b730525113b509d0f0a0419957d2a3880352a0109a23642113456ee250431209cf789980f0371b7050cb372274a24518cb6fbc5170711606672807846d7a33390a2d85d09de24631916c489d8804d905151458995a3dca6f85e21a23188c80f88674d0956c71c9ea81809dc3923ac1ae4778dfe618832b0893001c443a12adb5e4e8c340b03d5cb32d68c1299818da952c58083f6c20a8401c9c38084010c6864846397a7779f4d616465206f6e20746865206d6f6f6e20627920426c6f636b6e6174697665a0512ca7252eadd6ff1a82e0eab881afe9baae8f2f69afddfa5ffb3a227276e6cd8800000000000000008503074678d7",
  "log_entry_data": "f89b94a0b86991c6218b36c1d19d4a2e9eb0ce3606eb48f863a0ddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3efa0000000000000000000000000ef013a60f765f34b0fd7c5aaf83b9c65bb10a9afa000000000000000000000000023ddd3e3692d1861ed57ede224608875809e127fa0000000000000000000000000000000000000000000000000000000174876e800",
  "log_index": "0",
  "proof": [
    "f90131a0967c44262dfeff60c2af9aba0bece5cb031e5072f2978bcf7fb18360abad1d3ca0ebdf9d8b1746ad2d585fa5c657348250fa92fa0586d6c89ab40c186115bebc17a0feffed8388da8e4f4936717c660a42f34b93cce4583d0944198c27e043779f56a06f970da48b1447fb56f88d27d9b0d4ea8cf30a6e83b507734fd96e3a42c17f24a0748ff314a3f88b77e11ef6c7f716f58ac498ae3b3264d6b2e0bdb6c91030df1fa05f6b3f1d31dc2f30322199b97af434c1c40891648515c0c483c2518f981a10b1a06dfe85601a2d49321453acf4edeab412bdb950fc65b2337b042aa7972d4925eea01eb88ec9349d5f61858f5a14b90456b78fb3bb0121d90df936d9765a05e78031a047c4fa8e5283c0dbeae2e4b9e6250008551b02d2edce1dd87bf9ace4d687604b8080808080808080",
    "f851a08aa8e40f2823eb01a484cf576f2582137280289ca54eb3da743a36203ba963d0a023bad42a2491d2b478b82ff7e2d1b2beb569981a23c4376be21d1436c12d920a808080808080808080808080808080",
    "f8918080808080808080a0918f290dbb902b8bf23c852bd82c945a37d4d5d015e9e9485616f3fa6e12aabca092b063303221d5c76f441ae788fa44597bd71cf03d0c944993ce8428c5a661fda0a7fc9b765f5a541c286acb44fe3ccffd32c1585da784b88d9bd800d933e69826a0ea9675894b751d28c53159f2582481df16925392fef0a81e1a2baa2d9e08190b8080808080",
    "f90211a0abc02bfbddd2313e525124b626275a08e4d5437c837a114b11bd56a9421d04aaa0e71d7c39496e26fa67b5cc364a49e2e5701c8b5edbd5c58afc716e92ca9cc8f9a0cdec4707b811edbf0c81db09631c5a18dbfbda66d4793358257e5ec5c8993234a01356b944c23504dc1aa148f1a46d5e3c3f1b990a88b8ade4767d84aa7d4ca79fa06f883de36996ebd0b395f114e96a468c1aeaa287ca4f10e01648b0d60515a111a02ccaea122ce8542fd7ff9e02cec311f8d1b8ab7a0d949074e8f621765c58588ba01aeeca3be1c4e8e2581ed016d74442677645c0b423bb34e919ccf14009cbc92ba03bc2b368375a4de99317a01c548c6e564966cb317c3b4ba20c5cb0b181ac7c6ca04350da6c4a50dbb535c72bcd4632a5b94c15c709e2df6cd91269ce0faacd46c4a0743908223350d40bdb093c331dc0fbabe066f46440fdcf474f28acc48162c67da052aeea371b09dfe541719470a63a1fb12708c72eaa575bf723fea692a48b9d77a0aefd559fe69cb9163a4c3aa562e80beb3199afa446c449f048a38be05bc5fa00a05a373afc917a53eea199c158f7daea520dcc4d5b76c6cf87bb823b962f6cb506a0eb119ea50817395bd79be6e6fa748d64dc9ded81c63c6d61474210ab0003be7da0e4c36dd93a40a90993f083829469e0718aaf510e8eedf9ed5a7d8fdd4af92643a0aa1c5c12d65fdf25f74354991185dccecefb1e5813531ce8c9fdd2b2c7494a3c80",
    "f902cf20b902cb02f902c70183f2d2d4b9010000000000000000000000000000000000000000000000000000000000000000000800010000000000000002000100000000000000000000000000000000000000000200000000000008000008000000000000000000000000000000004000000000000000000000000000000000000000000000000000010000000010000000000000000000000000000000000000000200000000010000000000000000000000000000000000200080000000200000000000000000000000000000004000000000000002000000000000000000000000000000000000080000000000000000000000000000000000000000000200000004000000000000000000000000000000f901bcf89b94a0b86991c6218b36c1d19d4a2e9eb0ce3606eb48f863a0ddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3efa0000000000000000000000000ef013a60f765f34b0fd7c5aaf83b9c65bb10a9afa000000000000000000000000023ddd3e3692d1861ed57ede224608875809e127fa0000000000000000000000000000000000000000000000000000000174876e800f9011c9423ddd3e3692d1861ed57ede224608875809e127ff863a0dd85dc56b5b4da387bf69c28ec19b1d66e793e0d51b567882fa31dc50bbd32c5a0000000000000000000000000a0b86991c6218b36c1d19d4a2e9eb0ce3606eb48a0000000000000000000000000ef013a60f765f34b0fd7c5aaf83b9c65bb10a9afb8a0000000000000000000000000000000000000000000000000000000174876e8000000000000000000000000000000000000000000000000000000000000000040000000000000000000000000000000000000000000000000000000000000002f6175726f72613a343332456231663237333036363241443141393739314564333443423244424466373030313535350000000000000000000000000000000000",
  ],
  "receipt_data": "02f902c70183f2d2d4b9010000000000000000000000000000000000000000000000000000000000000000000800010000000000000002000100000000000000000000000000000000000000000200000000000008000008000000000000000000000000000000004000000000000000000000000000000000000000000000000000010000000010000000000000000000000000000000000000000200000000010000000000000000000000000000000000200080000000200000000000000000000000000000004000000000000002000000000000000000000000000000000000080000000000000000000000000000000000000000000200000004000000000000000000000000000000f901bcf89b94a0b86991c6218b36c1d19d4a2e9eb0ce3606eb48f863a0ddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3efa0000000000000000000000000ef013a60f765f34b0fd7c5aaf83b9c65bb10a9afa000000000000000000000000023ddd3e3692d1861ed57ede224608875809e127fa0000000000000000000000000000000000000000000000000000000174876e800f9011c9423ddd3e3692d1861ed57ede224608875809e127ff863a0dd85dc56b5b4da387bf69c28ec19b1d66e793e0d51b567882fa31dc50bbd32c5a0000000000000000000000000a0b86991c6218b36c1d19d4a2e9eb0ce3606eb48a0000000000000000000000000ef013a60f765f34b0fd7c5aaf83b9c65bb10a9afb8a0000000000000000000000000000000000000000000000000000000174876e8000000000000000000000000000000000000000000000000000000000000000040000000000000000000000000000000000000000000000000000000000000002f6175726f72613a343332456231663237333036363241443141393739314564333443423244424466373030313535350000000000000000000000000000000000",
  "receipt_index": "167",
}
`;
